File: showdown.py
import asyncio
import json
import time

import aiohttp
import websockets

import showdown_commands

class Showdown:

    def __init__(self, bot, id, pw, channel_id, voice_id, timeout=3600, ai=False):
        self.bot = bot
        self.id = id
        self.pw = pw
        self.channel_id = channel_id
        self.voice_id = voice_id
        self.timeout = timeout
        self.ai = ai

        self.ws = None
        self.timer = self.timeout
        self.bot_time = 0
        self.running = False
        
        self.listener = {} #channel id : on_message function
        self.rooms = {} #channel id : active room
        self.channels = {} #active room : channel id
        
        showdown_commands.load_commands(self.bot, self)
    
    def reset_time(self):
        self.timer = self.timeout
            
    async def sendprint(self, msg):
        print(msg)
        await self.ws.send(msg)
        
    async def check_timeout(self):
        while True:
            await asyncio.sleep(self.timer)
            self.timer = time.time() - self.bot_time
            if self.timer >= self.timeout:
                pass #timeout occurred #should return here
            else:
                self.timer = self.timeout - self.timer
        
    async def login(self, msg):
        login_msg = '|/trn ' + self.id + ',0,'
        
        async with aiohttp.ClientSession() as session:
            url = 'http://play.pokemonshowdown.com/action.php'
            data = {}
            data['act'] = 'login'
            data['name'] = self.id
            data['pass'] = self.pw
            data['challstr'] = msg[2] + '|' + msg[3]
            
            #assume no errors here (bad idea); login fails if nametaken with bad password, or bad challstr
            async with session.post(url, data=data) as response:
                s = json.loads((await response.text())[1:])
                login_msg += s['assertion']
        
        await self.ws.send(login_msg)
        await self.ws.send('|/avatar 27')
        
    async def handle_challenge(self, msg):
        dict = json.loads(msg[2])
        for user, format in dict['challengesFrom'].items():
            if user == 'psikh0':
                await self.ws.send('|/utm ' + 'null')
                await self.ws.send('|/accept psikh0')
        return
    
    #Q Up
    async def queue(self): pass
    
    #let showdown handle target legality
    async def choose(self, room, command, target, option=''):
        await self.ws.send(room + '|/choose ' + command + ' ' + target + option)
    
    #room responses always start with >ROOMID\n
    async def handle_room_response(self, response):
        r = response.split('\n')
        room_id = r[0][1:].rstrip()
        
        for line in r[1:]:
            msg = line.split('|')
            if len(msg) <= 1: continue
            
            response_type = msg[1]
            if response_type == 'init':
                if msg[2] == 'battle':
                    #only use this channel for now
                    self.rooms[self.channel_id] = room_id
                    self.channels[room_id] = self.channel_id
                    pass
                elif msg[2] == 'chat': pass
            else:
                pass
            
        return
    
    async def handle_global_response(self, response):
        r = response.split('\n')  
        for line in r:
            msg = line.split('|')
            if len(msg) <= 1: continue
            
            response_type = msg[1]
            if response_type == 'challstr':
                await self.login(msg) #TODO: if not logged in
            elif response_type == 'updatechallenges':
                await self.handle_challenge(msg)
            elif response_type == 'updatesearch':
                #games = json.loads(msg[2])['games']
                #games can just be a json 'null'
                #bind room / channel
                continue
            else:
                continue
        return
    
    async def handle_response(self, response):
        if response[0] == '>':
            return await self.handle_room_response(response)
        else:
            return await self.handle_global_response(response)
    
    async def connect_with_timeout(self, ctx):
        if self.ws is not None: return
        #TODO maybe: await asyncio.wait_for(ws.recv(), timeout=X)
        try:
            async with websockets.connect('ws://sim.smogon.com:8000/showdown/websocket') as self.ws:
                showdown_commands.load_showdown_commands(self.bot, self, ctx)
                async for response in self.ws:
                    print(response + ' ENDRESPONSE')
                    await self.handle_response(response)
        finally:
            self.ws = None
            showdown_commands.unload_showdown_commands(self.bot, self, ctx)
    
    async def close(self):
        if self.ws is None: return
        ws_temp, self.ws = self.ws, None
        await ws_temp.close()

File: showdown.js
"use strict";

const axios = require('axios');
const WebSocket = require('ws');

class ShowdownInstance {
    
    constructor(bot, config) {
        this.ws = null;
        this.bot = bot;
        this.name = config.showdown_username;
        this.pw = config.showdown_password;
        
        this.listeners = {}; //channel id : event_handler
        this.rooms = {}; // channel id: room
        this.channels = {}; // room : channel id
        this.queue = {}; //on game search, place channel in queue, on game found, shift and assign channel to game room
        
        this.channel_id = config.discord_channel_id; //temp
    }
    
    choose(room, action, target, op='') {
        this.ws.send(room + '|/choose ' + action + ' ' + target + op);
    }
    
    async login(r) {
        const data = {};
        data.act = 'login';
        data.name = this.name;
        data.pass = this.pw;
        data.challstr = r[2] + '|' + r[3];
        
        //assume post request does not fail for now;
        const res = await axios.post(ShowdownInstance.action_url, data);
        const info = JSON.parse(res.data.slice(1));
        const login_msg = '|/trn ' + this.name + ',0,' + info.assertion;
        this.ws.send (login_msg)
        this.ws.send('|/avatar 27');
    }
    
    handle_room_response(message) {
        const response = message.split('\n');
        const room = response.shift().slice(1).trim();
        
        for (const line of response) {
            const r = line.split('|');
            if (r.length <= 1) { continue; }
            
            const type = r[1];
            if (type === 'init') {
                //temp
                if (r[2] === 'battle') {
                    this.rooms[this.channel_id] = room;
                    this.channels[room] = this.channel_id;
                }
            }
        }
    }
    
    handle_global_response(message) {
        const response = message.split('\n');
        
        for (const line of response) {
            const r = line.split('|');
            if (r.length <= 1) { continue; }
            
            const type = r[1];
            if (type === 'challstr') {
                this.login(r); //TODO: handle promise rejection?
            } else if (type === 'updatechallenges') {
                //for testing purposes
                const d = JSON.parse(r[2]);
                for (const user in d['challengesFrom']) {
                    if (user === 'psikh0') {
                        this.ws.send('|/utm ' + 'null');
                        this.ws.send('|/accept pSikh0');
                    }
                }
            }
        }
    }
    
    handle_response(message) {
        if (message[0] === '>') {
            this.handle_room_response(message);
        } else {
            this.handle_global_response(message);
        }
    }
    
    connect(channel) {
        if (this.ws !== null) { return; }
        
        this.ws = new WebSocket(ShowdownInstance.url);
        
        this.ws.on('open', () => {
            console.log('connected to showdown!');
        });
        this.ws.on('error', (error) => {
            console.log('ERROR');
            this.cleanup(channel);
        });
        this.ws.on('close', (code, reason) => {
            console.log('CLOSE');
            this.cleanup(channel);
        });
        this.ws.on('message', (message) => {
            console.log(message); //assert msg is string ?
            this.handle_response(message)
        });
    }
    
    run(channel) {
        if (this.ws !== null) { return; } //temporary
        
        if (this.ws === null) { this.connect(channel); }
        this.add_listener(channel.id);
    }
    
    stop() {
        if (this.ws === null) { return; }
        this.ws.terminate(); //emits close
    }
    
    cleanup(channel) {
        this.remove_listener(channel.id);
        this.ws = null;
    }

    generate_commands(channel_id) { 
        const commands = (message) => {
            if (message.channel.id !== channel_id) { return; }
            if (!message.content.startsWith('~') || message.author.bot) { return; }
            
            const args = message.content.slice(1).toLowerCase().split(/\s+/);
            const action = args.shift();
            
            const room = this.rooms[channel_id];
            if (room === undefined) { return; } //unless command is to queue
            
            if (args.length < 1) { return; }
            const target = args[0];
            if (action === 'switch') {
                return this.choose(room, action, target);
            }
            if (action === 'move') {
                const op = (args[1] === 'mega' || args[1] === 'zmove') ? ' ' + args[1] :'';
                return this.choose(room, action, target, op);
            }
        };
        return commands;
    }
    
    add_listener(channel_id) {
        if (channel_id in this.listeners) { return; }
        const commands = this.generate_commands(channel_id);
        this.listeners[channel_id] = commands;
        this.bot.on('message', commands);
    }
    
    remove_listener(channel_id) {
        const commands = this.listeners[channel_id];
        if (commands === undefined) { return; }
        this.bot.off('message', commands);
        delete this.listeners[channel_id];
    }
}

ShowdownInstance.url = 'ws://sim.smogon.com:8000/showdown/websocket';
ShowdownInstance.action_url = 'https://play.pokemonshowdown.com/action.php';

const showdown_commands = function(bot, config) {
    
    const pokemon = new ShowdownInstance(bot, config);
    
    const showdown = {
        name: 'showdown',
        execute: function(message, args) {
            if (message.channel.id !== config.discord_channel_id) { return; } //block calls from other channels for now
            pokemon.run(message.channel);
        },
    };
    
    const command_list = {
        showdown: showdown,
    };
    
    const stop = () => {
        pokemon.stop();
    }
    
    return {
        command_list: command_list,
        stop: stop,
    };
};

module.exports = showdown_commands;